adaptdl/adaptdl/checkpoint.py

```diff
@@ -95,24 +99,41 @@ def save_all_states():
     for state in _STATES_TO_NAMES:
         save_state(state)
 
+    # Prevent corrupting original state files in case the process got killed
+    # during state file writing.
+    if replica_rank() == 0 and checkpoint_path() is not None:
+        ckpt_dir = os.path.join(checkpoint_path(), CKPT_DIR_NAME)
+        if os.path.isdir(ckpt_dir):
+            shutil.rmtree(ckpt_dir)
+
+        os.makedirs(ckpt_dir, exist_ok=True)
+        ckpt_tmp_dir = os.path.join(checkpoint_path(), CKPT_TMP_DIR_NAME)
+        os.rename(ckpt_tmp_dir, ckpt_dir)  # atomic
```

Comment:
The process is still not quite atomic. e.g. say the pod gets killed during the `rmtree` (which can take a non-trivial amount of time when using a DFS). A better way is to save to a unique directory each time, e.g. `checkpoint-{n}` where `n = adaptdl.env.num_restarts()`, and use a symlink to point to the latest directory. Then, delete the old checkpoints only after the symlink is written.

Reply:
You are right. I will make the change as suggested. Thanks!